compiler/src/liveness.ml

```diff
@@ -50,7 +50,14 @@ and live_d weak d (s_o: Sv.t) =
     let s2, c2 = live_c weak c2 s_o in
     Sv.union (vars_e e) (Sv.union s1 s2), s_o, Cif(e, c1, c2)
 
-  | Cfor _ -> assert false (* Should have been removed before *)
+  | Cfor (x, r, c) ->
+    let rec loop s_o =
+      let s_i, c = live_c weak c s_o in
+      let s_i = Sv.remove (L.unloc x) s_i in
+      if Sv.subset s_i s_o then s_i, c
+      else loop (Sv.union s_i s_o) in
+    let s_i, c = loop s_o in
+    s_i, (if weak then Sv.add (L.unloc x) s_o else s_o), Cfor (x, r, c)
```

Comment:
Why do you look at `weak`?

The variables occurring in the range are live before the `for` instruction.